File: 백준/Gold/9251. LCS/LCS.py
X = input()
Y = input()

n = len(X)
m = len(Y)

X = '0' + X
Y = '0' + Y

dp_table = [[0] * (m+1) for _ in range(n+1)]

for i in range(1, n+1):
    for j in range(1, m+1):
        if X[i] == Y[j]:
            dp_table[i][j] = dp_table[i-1][j-1] + 1
        else:
            dp_table[i][j] = max(dp_table[i][j-1], dp_table[i-1][j])

print(dp_table[n][m])

File: 백준/Gold/9251. LCS/LCS.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        String X = sc.nextLine();
        String Y = sc.nextLine();

        int n = X.length();
        int m = Y.length();

        X = 'x' + X;
        Y = 'y' + Y;

        int[][] dpTable = new int[n + 1][m + 1];

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (X.charAt(i) == Y.charAt(j)) {
                    dpTable[i][j] = dpTable[i - 1][j - 1] + 1;
                } else {
                    dpTable[i][j] = Math.max(dpTable[i][j - 1], dpTable[i - 1][j]);
                }
            }
        }

        System.out.println(dpTable[n][m]);

        sc.close();
    }
}
